### PR TITLE
UCT: wakeup refactoring

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe
Wakeup is a part of interface, there is no need for a separate handle

Added API definition to receive a wakeup event via callback

UCP and tests are updated.

### CODE REVIEW

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -927,41 +917,52 @@ ucs_status_t uct_iface_get_address(uct_iface_h iface, uct_iface_addr_t *addr);
...
- * @param [in]  wakeup     Handle to the notification event.
- * @param [out] fd_p       Location to write the notification file descriptor.
+ * @param [in]  iface      Interface to set the notification callback
+ * @param [in]  cb         Notification callback. If NULL the previous callback is 
+ *                         removed. See @uct_wakeup_cb_t
  *
  * @return Error code.
  */
-ucs_status_t uct_wakeup_efd_get(uct_wakeup_h wakeup, int *fd_p);
+ucs_status_t uct_iface_wakeup_cb_set(uct_iface_h iface, uct_wakeup_cb_t cb);
```
Comment (yosefe) [inline src/uct/api/uct.h]: should be passed in uct_iface_params

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -982,29 +984,30 @@ ucs_status_t uct_wakeup_efd_get(uct_wakeup_h wakeup, int *fd_p);
...
  * @ingroup UCT_RESOURCE
  * @brief Cause the next notification.
  *
- * @param [in] wakeup      Handle to the notification event.
+ * @param [in] iface      Interface to notify 
  *
  * @return Error code.
  */
-ucs_status_t uct_wakeup_signal(uct_wakeup_h wakeup);
+ucs_status_t uct_iface_signal(uct_iface_h iface);
```
Comment (yosefe) [inline src/uct/api/uct.h]: uct_iface_wakeup_signal

File: src/uct/base/uct_iface.c [hunk 3]
```diff
@@ -160,49 +160,24 @@ int uct_iface_is_reachable(const uct_iface_h iface, const uct_device_addr_t *dev
...
-
-ucs_status_t uct_wakeup_efd_get(uct_wakeup_h wakeup, int *fd_p)
-{
-    return wakeup->iface->ops.iface_wakeup_get_fd(wakeup, fd_p);
-}
-
-ucs_status_t uct_wakeup_efd_arm(uct_wakeup_h wakeup)
-{
-    return wakeup->iface->ops.iface_wakeup_arm(wakeup);
+    return iface->ops.iface_wakeup_get_fd(iface, fd_p);
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: pls rename func pointer to iface->ops.iface_wakeup_efd_get

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -927,41 +917,52 @@ ucs_status_t uct_iface_get_address(uct_iface_h iface, uct_iface_addr_t *addr);
 int uct_iface_is_reachable(const uct_iface_h iface, const uct_device_addr_t *dev_addr,
                            const uct_iface_addr_t *iface_addr);
 
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief  Wakeup event types.
+ */
+enum uct_wakeup_event_types {
+    UCT_WAKEUP_TX_COMPLETION   = UCS_BIT(0),
```
Comment (yosefe) [inline src/uct/api/uct.h]: pls set following events. we need also signaled tx completion to allow specific signaling for protocols such as rendezvous.
```
UCT_WAKEUP_TX_COMP,
UCT_WAKEUP_TX_SIGNALED_COMP,
UCT_WAKEUP_RX_AM,
UCT_WAKEUP_RX_SIGNALED_AM,
UCT_WAKEUP_TX
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1188/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3133/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1192/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3137/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1196/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3142/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3906/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.c [hunk 5]
```diff
@@ -137,93 +134,104 @@ static void ucp_worker_am_tracer(void *arg, uct_am_trace_type_t type,
...
-    return uct_events;
+    return UCS_OK;
+
+err_pipe_cleanup:
+    close(worker->wakeup_pipe[0]);
+    close(worker->wakeup_pipe[1]);
+err_close_epfd:
+    close(worker->epfd);
+out:
+    return UCS_OK;
```
Comment (brminich) [inline src/ucp/core/ucp_worker.c]: return status

File: src/uct/api/uct.h [hunk 6]
```diff
@@ -980,38 +979,17 @@ ucs_status_t uct_ep_check(const uct_ep_h ep, unsigned flags,
...
- */
-void uct_wakeup_close(uct_wakeup_h wakeup);
-
-
-/**
- * @ingroup UCT_RESOURCE
  * @brief Obtain a notification file descriptor for polling.
  *
- * @param [in]  wakeup     Handle to the notification event.
+ * Only interfaces with the @ref UCT_IFACE_FLAG_WAKEUP_FD
```
Comment (brminich) [inline src/uct/api/uct.h]: UCT_IFACE_FLAG_WAKEUP_FD -> UCT_IFACE_FLAG_EVENT_FD

File: src/uct/api/uct.h [hunk 7]
```diff
@@ -1135,7 +1092,7 @@ void uct_iface_mem_free(const uct_allocated_memory_t *mem);
  * @param [in]  id       Active message id. Must be 0..UCT_AM_ID_MAX-1.
  * @param [in]  cb       Active message callback. NULL to clear.
  * @param [in]  arg      Active message argument.
- * @param [in]  flags    Required @ref uct_am_cb_cap "active message callback capabilities"
+ * @param [in]  flags    Required @ref uct_am_cb_flags_t "active message callback flags"
```
Comment (brminich) [inline src/uct/api/uct.h]: It is unrelated, right? We have uct_am_cb_flags (without _t postfix).

File: src/uct/api/uct_def.h [hunk 8]
```diff
@@ -347,4 +346,13 @@ typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, unsigned flags,
                                                 const void *rkey_buf);
 
 
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief  Interface event callback.
+ *
+ * @param [in]  arg      User argument to be passed to the callback.
+ */
+typedef void (*uct_iface_event_cb_t)(void *arg);
```
Comment (brminich) [inline src/uct/api/uct_def.h]: How will it be used?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3914/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3920/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3923/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by yosefe** [comment]
@shamisp @bbenton pls tak a look

**Comment by yosefe** [comment]
@MattBBaker can you pls post stacktrace of ORNL hang?

**Comment by MattBBaker** [comment]
The current hang up looks like a jenkins problem. The build before the most recent one looked successful.

**Comment by shamisp** [comment]
Looks good.
Adding @bbenton

File: src/uct/api/uct.h [hunk 9]
```diff
@@ -980,38 +979,16 @@ ucs_status_t uct_ep_check(const uct_ep_h ep, unsigned flags,
...
- */
-void uct_wakeup_close(uct_wakeup_h wakeup);
-
-
-/**
- * @ingroup UCT_RESOURCE
  * @brief Obtain a notification file descriptor for polling.
  *
- * @param [in]  wakeup     Handle to the notification event.
+ * Only interfaces with the @ref UCT_IFACE_FLAG_EVENT_FD implement this function.
```
Comment (bbenton) [inline src/uct/api/uct.h]: wording suggestion:
s/Only interfaces with the/Only interfaces supporting/

File: src/uct/api/uct.h [hunk 10]
```diff
@@ -1020,9 +997,10 @@ ucs_status_t uct_wakeup_efd_get(uct_wakeup_h wakeup, int *fd_p);
  *
  * This routine needs to be called before waiting on each notification on this
  * interface, so will typically be called once the processing of the previous
- * event is over, as part of the wake-up mechanism.
+ * event is over.
  *
- * @param [in] wakeup      Handle to the notification event.
+ * @param [in] iface       Interface to arm.
+ * @param [in] events      Events to wakeup on. See @ref uct_iface_event_types_t
```
Comment (bbenton) [inline src/uct/api/uct.h]: fix reference:
s/uct_iface_event_types_t/uct_iface_event_types/

**Comment by yosefe** [comment]
@bbenton fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3950/ for details (Mellanox internal link).

**Comment by bbenton** [comment]
:+1: for Doc/API

**Pull request merged**